### PR TITLE
Simplified Landing Zone creation in `sodar itransfer-*`

### PR DESCRIPTION
**Pull request by eudesbarbosa**
**Context** 

While using `sodar itransfer-*` users would have to provide an iRODS path or UUID of previously (manually) created Landing Zone.

**Goal** 

Expand this functionality to allow users to automatically list available Landing Zones or create a new one. Use cases:

1. User provides iRODS path. Same as before, use it.
2. User provides Landing Zone UUID. Same as before, fetch path and use it.
3. User provides Project UUID:
    i. If there are LZ associated with project, select the latest active and use it.
    ii. If there are no LZ associated with project, create a new one and use it.

Step **3** can be done automatically if new flag `--yes` was used, or by asking user for input.

**Changes** 

The following methods were created in `SnappyItransferCommandBase`:
* `get_sodar_info()`
* `move_landing_zone()`
* `get_landing_zone_by_uuid()`
* `create_landing_zone()`
* `get_latest_landing_zone()`

The following flags were added to `SnappyItransferCommandBase`:
* `--yes`: Assume all answers are yes, e.g., will create or use existing available landing zones without asking,
* `--validate-and-move`: fter files are transferred to SODAR, it will proceed with validation and move them.

**Tests**
1. There were no additional error nor warning messages while using the automatic `pytest`.
2. Passed manual tests in sandbox tests ([logs](https://github.com/bihealth/cubi-tk/files/5944631/8_cubi-tk_tests.log)).

### CODE REVIEW

File: cubi_tk/snappy/itransfer_common.py [hunk 1]
```diff
@@ -329,7 +332,190 @@ def build_jobs(self, library_names) -> typing.Tuple[TransferJob, ...]:
...
+
+    def get_latest_landing_zone(self, project_uuid):
+        """
+        :param project_uuid: Project UUID.
+        :type project_uuid: str
+
+        :return: Returns landing zone UUID and iRODS path in latest active landing zone available.
+        If none available, it returns None for both.
+        """
+        #
```
Comment (holtgrewe) [inline cubi_tk/snappy/itransfer_common.py:493]: why the empty comment line?

File: cubi_tk/snappy/itransfer_common.py [hunk 2]
```diff
@@ -329,7 +332,190 @@ def build_jobs(self, library_names) -> typing.Tuple[TransferJob, ...]:
...
+        from ..sodar.api import landing_zones
```
Comment (holtgrewe) [inline cubi_tk/snappy/itransfer_common.py:494]: can we move the imports to be global?

**Review by holtgrewe — CHANGES_REQUESTED** [review]

File: cubi_tk/snappy/itransfer_common.py [hunk 3]
```diff
@@ -329,7 +332,190 @@ def build_jobs(self, library_names) -> typing.Tuple[TransferJob, ...]:
...
+
+    def move_landing_zone(self, lz_uuid):
+        """
+        Method calls SODAR API to validate and move transferred files.
+
+        :param lz_uuid: Landing zone UUID.
+        :type lz_uuid: str
+        """
+        from ..sodar.api import landing_zones
+        logger.info("Transferred files move to Landing Zone {uuid} will "
```
Comment (holtgrewe) [inline cubi_tk/snappy/itransfer_common.py:444]: This looks like a bug, why not `logger.info("xxx %s" " xxx", lz_uuid)`?

**Review by holtgrewe — CHANGES_REQUESTED** [review]

File: cubi_tk/snappy/itransfer_common.py [hunk 4]
```diff
@@ -329,7 +332,190 @@ def build_jobs(self, library_names) -> typing.Tuple[TransferJob, ...]:
...
+
+            if create_lz_bool:
+                # Assume that provided UUID is associated with a Project and user wants a new LZ.
+                # Behavior: create new LZ.
+                try:
+                    lz_uuid, lz_irods_path = self.create_landing_zone(project_uuid=in_destination)
+                except requests.exceptions.HTTPError as e:
+                    logger.error("Unable to create Landing Zone using UUID {uuid}. "
+                                 "HTTP error {error}".format(uuid=in_destination, error=str(e)))
+                    exit(os.EX_DATAERR)
```
Comment (holtgrewe) [inline cubi_tk/snappy/itransfer_common.py:373]: The pythonic way is to use exceptions here. Look at the existing `cubi_tk.exceptions` module.

File: cubi_tk/snappy/itransfer_common.py [hunk 5]
```diff
@@ -281,17 +292,9 @@ def build_base_dir_glob_pattern(
 
     def build_jobs(self, library_names) -> typing.Tuple[TransferJob, ...]:
```
Comment (holtgrewe) [inline cubi_tk/snappy/itransfer_common.py:293]: Please adjust the return type here also.

File: cubi_tk/snappy/itransfer_common.py (hunk 4 above)
Comment (eudesbarbosa) [reply cubi_tk/snappy/itransfer_common.py]: Changed.

File: cubi_tk/snappy/itransfer_common.py [hunk 6]
```diff
@@ -330,7 +333,189 @@ def build_jobs(self, library_names) -> typing.Tuple[TransferJob, ...]:
...
+                        logger.info(f"Found active Landing Zone: {lz_irods_path}")
+                        if not input("Can the process use this path? [yN] ").lower().startswith("y"):
+                            logger.info(f"...an alternative is to create another Landing Zone "
+                                        f"using the UUID {in_destination}")
+                            if input("Can the process create a new landing zone? [yN] ").lower().startswith("y"):
+                                lz_uuid, lz_irods_path = self.create_landing_zone(project_uuid=in_destination)
+                            else:
+                                logger.info("Not possible to continue the process without a "
+                                            "landing zone path. Breaking...")
+                                raise UserCanceledException
```
Comment (holtgrewe) [inline cubi_tk/snappy/itransfer_common.py:412]: raise object not class

File: cubi_tk/snappy/itransfer_common.py [hunk 7]
```diff
@@ -330,7 +333,189 @@ def build_jobs(self, library_names) -> typing.Tuple[TransferJob, ...]:
...
+            if create_lz_bool:
+                # Assume that provided UUID is associated with a Project and user wants a new LZ.
+                # Behavior: create new LZ.
+                try:
+                    lz_uuid, lz_irods_path = self.create_landing_zone(project_uuid=in_destination)
+                except requests.exceptions.HTTPError as e:
+                    exception_str = str(e)
+                    logger.error(f"Unable to create Landing Zone using UUID {in_destination}. "
+                                 f"HTTP error {exception_str}")
+                    raise e
```
Comment (holtgrewe) [inline cubi_tk/snappy/itransfer_common.py:374]: Proposal: shorten `raise e` to `raise`

**Review by holtgrewe — CHANGES_REQUESTED** [review]

**Review by holtgrewe — APPROVED** [review]

**Pull request merged**